### PR TITLE
feat: use lua api for resizing

### LINKED ISSUE #111: Autoresize set to false is not working.
Hi and first of all thank you for the amazing plugin.

With the new update the autoresize option does't do anything.
In the settings i have set `autoresize = { { enable = false} }` and it doesn't work.

https://github.com/nvim-focus/focus.nvim/assets/39212564/8b3d0249-d8e7-42e7-842b-176d96dcde58

Because of this, the ignored filetypes, like nvim-tree or trouble.nvim are't ignored by focus.

https://github.com/nvim-focus/focus.nvim/assets/39212564/481e4269-cc8f-4537-8e53-10763617d71f

Here is my configuration:

```lua
return {
    'nvim-focus/focus.nvim',
    enabled = true,
    event = 'VeryLazy',
    version = '*',
    keys = {
        { '<space>h', '<cmd>FocusSplitLeft<CR>', desc = 'Focus Split Left' },
        { '<space>j', '<cmd>FocusSplitDown<CR>', desc = 'Focus Split Down' },
        { '<space>k', '<cmd>FocusSplitUp<CR>', desc = 'Focus Split Up' },
        { '<space>l', '<cmd>FocusSplitRight<CR>', desc = 'Focus Split Right' },
        { '<space>m', '<cmd>FocusMaxOrEqual<CR>', desc = 'Maximaze or Equalises Split' },
    },
    config = function()
        require('focus').setup({
            autoresize = {
                enable = false,
            },
            ui = {
                number = false,
                relativenumber = false,
                hybridnumber = false,
                signcolumn = false,
            },
            -- excluded_filetypes = {
            --     'toggleterm',
            --     'qf',
            --     'help',
            --     'NvimTree',
            --     'Trouble',
            --     'tsplayground',
            --     'Outline',
            -- },
        })

        -- Ignore filetypes
        local ignore_filetypes = {
            'toggleterm',
            'qf',
            'help',
            'NvimTree',
            'Trouble',
            'tsplayground',
            'Outline',
        }
        -- Ignore buftypes
        local ignore_buftypes = {
            'nofile',
            'prompt',
            'popup',
        }

        local augroup = vim.api.nvim_create_augroup('FocusDisable', { clear = true })
        vim.api.nvim_create_autocmd('WinEnter', {
            group = augroup,
            callback = function(_)
                if vim.tbl_contains(ignore_buftypes, vim.bo.buftype) then
                    vim.b.focus_disable = true
                end
            end,
            desc = 'Disable focus autoresize for BufType',
        })

        vim.api.nvim_create_autocmd('FileType', {
            group = augroup,
            callback = function(_)
                if vim.tbl_contains(ignore_filetypes, vim.bo.filetype) then
                    vim.b.focus_disable = true
                end
            end,
            desc = 'Disable focus autoresize for FileType',
        })
    end,
}
```
My plugin manager is lazy.nvim.

The `excluded_filetype` option like before this update, was in my opinion much cleaner. It would be nice if you bring it back. 🙂

Thank you again.

### PR DESCRIPTION
**Pull request by willothy**
This PR is now just a refactor using the Lua API for resizing instead of window options, plus a few fixes and features that came with that. 

The animation feature has been removed since it was a bit out of scope for this PR.

closes #111

### CODE REVIEW

File: lua/focus/modules/resizer.lua [hunk 1]
```diff
@@ -21,76 +21,180 @@ local golden_ratio_minheight = function()
     return math.floor(golden_ratio_height() / (3 * golden_ratio))
 end
 
+local easing = {}
+
+function easing.linear(t, b, c, d)
+    return c * t / d + b
+end
+
```
Comment (willothy) [inline lua/focus/modules/resizer.lua:29]: These won't all be here, just testing what looks best and I'm planning on choosing 3 or 4 to have as options.

File: lua/focus/modules/resizer.lua [hunk 2]
```diff
@@ -21,76 +21,312 @@ local golden_ratio_minheight = function()
...
-                        .. 'set minheight to at least '
-                        .. golden_ratio_height()
-                )
+        vim.api.nvim_win_set_width(win, width)
+        vim.api.nvim_win_set_height(win, height)
+        vim.fn.winrestview(view)
+    end
+end
+
+function M.equalise(config)
```
Comment (willothy) [inline lua/focus/modules/resizer.lua:233]: This has some issues still, struggles with odd numbers of windows

**Comment by cryptomilk** [comment]
Don't forget to add tests ;-)

**Comment by willothy** [comment]
> Don't forget to add tests ;-)

Of course! I'm just finishing up the equalize code, it works without animation but gets thrown off by the recursive window updates when animating. Will add tests once the feature is working properly :)

**Comment by willothy** [comment]
Here's a demo of the animated smart equalizer - still some interpolation / rounding issues to solve but it's getting there:


https://github.com/nvim-focus/focus.nvim/assets/38540736/181715a5-a24a-4ce3-8fa4-0843dccea255

**Comment by willothy** [comment]
Honestly I'm not very happy with how messy the code in this PR has turned out. I do want to implement the feature, I just think that the approach I took was a bit too naive. I'm working on a managed-state window system that should handle any resizing strategy better than only tracking the current window can. I'll merge that into this branch when it's done :)

**Comment by willothy** [comment]
I'm not totally sure I want to move forward with this PR. I can't seem to get older tests to pass, and I think the plugin is probably better off without this feature for now.

**Comment by willothy** [comment]
May revisit this PR later, but gonna close it for now.

**Comment by cryptomilk** [comment]
That's fine. Thank for your efforts!

Lets cleanup the current codebase first :-) If you have time, could you look into #117 ? I don't have time right now.

**Comment by willothy** [comment]
> Lets cleanup the current codebase first :-) If you have time, could you look into #117 ? I don't have time right now.

Agreed! And totally, I should have some time to look into it in the next few days.

**Comment by willothy** [comment]
Tentatively reopening this without the animation feature as I think using the nvim api instead of window options may solve some current issues and give us more granular control over resizing.

File: README.md [hunk 3]
```diff
@@ -13,19 +13,19 @@ Always have a nice view over your split windows
 
 ![screencast](https://i.ibb.co/0tsKww4/focusop.gif)
 
-*Note*: For reference this screencast features dimensions set to 40 rows and 120
+_Note_: For reference this screencast features dimensions set to 40 rows and 120
 columns.
```
Comment (willothy) [inline README.md:17]: Sorry about this, autoformat got me. I'll revert readme.md

**Comment by willothy** [comment]
@cryptomilk I think this is pretty much ready :)

**Comment by cryptomilk** [comment]
Thanks, I will test it. Why are you removing several split tests?

**Comment by willothy** [comment]
> Thanks, I will test it. Why are you removing several split tests?

I removed tests that asserted window options, they were failing because those window options aren't used for resizing anymore. The tests that assert actual window size still pass.

**Comment by cryptomilk** [comment]
Please add a comment to the commit message explaining this.

Also the commit changing `vim.b.focus_disable` to `vim.w.focus_disable` should be marked with a `!` (breaking change)`. This will be probably displayed in lazy as a breaking change then.

See https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#commit-messages

File: lua/focus/modules/functions.lua [hunk 4]
```diff
@@ -8,6 +8,7 @@ M.focus_enable = function()
     end
 
     vim.g.focus_disable = false
+    require('focus.modules.resizer').goal = 'autoresize'
```
Comment (cryptomilk) [inline lua/focus/modules/functions.lua:11]: Can't we pass that down as an argument to the `resize()` function?

Comment (cryptomilk) [reply lua/focus/modules/functions.lua:11]: Also rename `goal` to `action` and have the default action set to `autoresize`?

**Comment by willothy** [comment]
> Please add a comment to the commit message explaining this.
> 
> Also the commit changing `vim.b.focus_disable` to `vim.w.focus_disable` should be marked with a `!` (breaking change)`. This will be probably displayed in lazy as a breaking change then.

Gotcha, I squashed that into the second line of another commit's message, but I'll add a breaking notice to it.

File: lua/focus/modules/functions.lua (hunk 4 above)
Comment (willothy) [reply lua/focus/modules/functions.lua:11]: good idea! I'll make that change now.

**Comment by willothy** [comment]
Also added disabling by filetype, as it's more friendly to new users to provide a config option.

**Comment by cryptomilk** [comment]
Please put the commit for filetype handling in its own MR. This needs further discussion.

**Comment by willothy** [comment]
> Please put the commit for filetype handling in its own MR. This needs further discussion.

Will do. I'll instead update the readme to tell users to use BufWinEnter and window-local vars.

File: lua/focus/modules/resizer.lua [hunk 5]
```diff
@@ -21,76 +21,82 @@ local golden_ratio_minheight = function()
...
-                        .. 'golden_ratio_height derived from your display. Please '
-                        .. 'set minheight to at least '
-                        .. golden_ratio_height()
-                )
-            else
-                vim.o.winminheight = config.autoresize.minheight
-            end
-        else
-            vim.o.winminheight = golden_ratio_minheight()
+            height = math.max(height, config.autoresize.minheight)
```
[the remark this answers is not in the document]
Comment (willothy) [reply lua/focus/modules/resizer.lua:43]: Added tests back and fixed min width/height

**Comment by cryptomilk** [comment]
Thanks, looks fine for me now.

**Review by cryptomilk — APPROVED** [review]
lgtm

**Pull request merged**